src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c

```diff
@@ -0,0 +1,393 @@
+  SODERLIND_K3(C) = k3;
+  SODERLIND_K4(C) = k4;
+  SODERLIND_K5(C) = k5;
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+/* -----------------------------------------------------------------
+ * Function to create a PID controller (subset of Soderlind)
+ */
+
+SUNAdaptController SUNAdaptController_PID(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
+  (void) SUNAdaptController_SetParams_PID(C,
+                                          DEFAULT_PID_K1,
+                                          DEFAULT_PID_K2,
+                                          DEFAULT_PID_K3);
+  return (C);
+}
+
+/* -----------------------------------------------------------------
+ * Function to set PID parameters
+ */
+
+int SUNAdaptController_SetParams_PID(SUNAdaptController C,
+                                     sunrealtype k1, sunrealtype k2,
+                                     sunrealtype k3)
+{
+  SODERLIND_K1(C) = k1;
+  SODERLIND_K2(C) = k2;
+  SODERLIND_K3(C) = k3;
+  SODERLIND_K4(C) = RCONST(0.0);
+  SODERLIND_K5(C) = RCONST(0.0);
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+/* -----------------------------------------------------------------
+ * Function to create a PI controller (subset of Soderlind)
+ */
+
+SUNAdaptController SUNAdaptController_PI(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
+  (void) SUNAdaptController_SetParams_PI(C,
+                                         DEFAULT_PI_K1,
+                                         DEFAULT_PI_K2);
+  return (C);
+}
+
+/* -----------------------------------------------------------------
+ * Function to set PI parameters
+ */
+
+int SUNAdaptController_SetParams_PI(SUNAdaptController C,
+                                    sunrealtype k1, sunrealtype k2)
+{
+  SODERLIND_K1(C) = k1;
+  SODERLIND_K2(C) = k2;
+  SODERLIND_K3(C) = RCONST(0.0);
+  SODERLIND_K4(C) = RCONST(0.0);
+  SODERLIND_K5(C) = RCONST(0.0);
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+/* -----------------------------------------------------------------
+ * Function to create an I controller (subset of Soderlind)
+ */
+
+SUNAdaptController SUNAdaptController_I(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
+  (void) SUNAdaptController_SetParams_I(C, DEFAULT_I_K1);
+  return (C);
+}
+
+/* -----------------------------------------------------------------
+ * Function to set PI parameters
+ */
+
+int SUNAdaptController_SetParams_I(SUNAdaptController C, sunrealtype k1)
+{
+  SODERLIND_K1(C) = k1;
+  SODERLIND_K2(C) = RCONST(0.0);
+  SODERLIND_K3(C) = RCONST(0.0);
+  SODERLIND_K4(C) = RCONST(0.0);
+  SODERLIND_K5(C) = RCONST(0.0);
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+/* -----------------------------------------------------------------
+ * Function to create an explicit Gustafsson controller
+ */
+
+SUNAdaptController SUNAdaptController_ExpGus(SUNContext sunctx)
+{
+  SUNAdaptController C;
```

Comment:
```suggestion
  SUNAdaptController C;
  if (sunctx == NULL) { return NULL; }
```

Reply:
It is redundant but I don't think we should rely on a call to another function for input checks.